stdlib/public/SDK/Foundation/DataProtocol.swift

```diff
@@ -99,25 +99,47 @@ extension DataProtocol {
     
     @discardableResult
     public func copyBytes<R: RangeExpression>(to ptr: UnsafeMutableRawBufferPointer, from range: R) -> Int where R.Bound == Index {
-        let bound = ptr.bindMemory(to: UInt8.self)
-        return copyBytes(to: bound, from: range)
-    }
-    
-    @discardableResult
-    public func copyBytes<DestinationType, R: RangeExpression>(to ptr: UnsafeMutableBufferPointer<DestinationType>, from range: R) -> Int where R.Bound == Index {
         precondition(ptr.baseAddress != nil)
 
         let concreteRange = range.relative(to: self)
         let slice = self[concreteRange]
 
-        return ptr.withMemoryRebound(to: UInt8.self) { buffer -> Int in
-            var (remaining, idx) = slice._copyContents(initializing: buffer)
-            while let next = remaining.next(), idx < ptr.endIndex {
-                buffer[idx] = next
-                idx += 1
+        // For contiguous types, we should be able to access an underlying buffer through Sequence.withUnsafeBytesIfSupported.
+        let amountCopied = slice.withUnsafeBytesIfSupported { buffer -> Int in
```

Comment:
Thanks! To be clear the `withMemoryRebound` call was safe in this case, but it is a signal that something's wrong with the API. In this case, the problem was that `_copyContents` is not supposed to be a raw "memcopy". I think the fundamental problem is that we don't have a way to constrain a protocol extension on something like "Element : TrivialType". If we did, then we would have a variant of `_copyContents` on trivial `RandomAccessCollections` taking an `UnsafeMutableRawBufferPointer` as the copy destination.

IMO, it's better to duplicate some code, as you've done, rather than abuse the existing API.